modules/rest/src/main/java/org/apache/ignite/internal/rest/RestComponent.java

```diff
@@ -17,151 +17,161 @@
 
-    /** Node configuration register. */
-    private final ConfigurationRegistry nodeCfgRegistry;
+    /** Factories that produce beans needed for REST controllers. */
+    private final List<RestFactory> restFactories;
 
-    /** Netty bootstrap factory. */
-    private final NettyBootstrapFactory bootstrapFactory;
+    private final RestConfiguration restConfiguration;
 
-    private final SimpleRouter router = new SimpleRouter();
+    /** Micronaut application context. */
+    private volatile ApplicationContext context;
 
-    /** Netty channel. */
-    private volatile Channel channel;
+    /** Server port. */
+    private int port;
 
     /**
      * Creates a new instance of REST module.
-     *
-     * @param nodeCfgMgr       Node configuration manager.
-     * @param bootstrapFactory Netty bootstrap factory.
      */
-    public RestComponent(ConfigurationManager nodeCfgMgr, NettyBootstrapFactory bootstrapFactory) {
-        nodeCfgRegistry = nodeCfgMgr.configurationRegistry();
-
-        this.bootstrapFactory = bootstrapFactory;
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public void registerHandlers(Consumer<Routes> registerAction) {
-        registerAction.accept(router);
+    public RestComponent(List<RestFactory> restFactories, RestConfiguration restConfiguration) {
+        this.restFactories = restFactories;
+        this.restConfiguration = restConfiguration;
     }
 
     /** {@inheritDoc} */
     @Override
     public void start() {
-        if (channel != null) {
-            throw new IgniteException("RestModule is already started.");
-        }
-
-        channel = startRestEndpoint(router).channel();
-    }
-
-    /**
-     * Start endpoint.
-     *
-     * @param router Dispatcher of http requests.
-     * @return Future which will be notified when this channel is closed.
-     * @throws RuntimeException if this module cannot be bound to a port.
-     */
-    private ChannelFuture startRestEndpoint(Router router) {
-        RestView restConfigurationView = nodeCfgRegistry.getConfiguration(RestConfiguration.KEY).value();
+        RestView restConfigurationView = restConfiguration.value();
 
         int desiredPort = restConfigurationView.port();
         int portRange = restConfigurationView.portRange();
 
-        int port = 0;
-        Channel ch = null;
-
-        ServerBootstrap bootstrap = bootstrapFactory.createServerBootstrap()
-                .handler(new LoggingHandler(LogLevel.INFO))
-                .childHandler(new RestApiInitializer(router));
-
         for (int portCandidate = desiredPort; portCandidate <= desiredPort + portRange; portCandidate++) {
-            ChannelFuture bindRes = bootstrap.bind(portCandidate).awaitUninterruptibly();
-
-            if (bindRes.isSuccess()) {
-                ch = bindRes.channel();
+            try {
+                context = buildMicronautContext(portCandidate).start();
                 port = portCandidate;
```

Comment:
unfortunately this is not correct. If you want to avoid making `port` variable `volatile` and guard it with `context` accesses, you must always write to the `port` variable **before** writing to the `context` variable. Same applies to reading, but in reverse order

Reply:
Thanks, you are absolutelly right.